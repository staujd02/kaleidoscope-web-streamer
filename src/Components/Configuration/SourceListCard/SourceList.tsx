import React from 'react';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';

export default class SourceList extends React.Component<SourceListProps, SourceListState> {

    constructor(props: SourceListProps) {
        super(props);
        this.state = {
            selectedIndex: -1
        }
    }

    render() {
        const { selectedIndex } = this.state;
        const { search } = this.props;
        const sourceList = this.props.sources
            .filter(source => search ? source.title.includes(search) 
            || source.source.includes(search) : true)
            .sort((a, b) => a.sortOrder - b.sortOrder)
            .map((source, idx) => (
                <ListItem
                    key={idx}
                    button
                    className={!source.isEnabled ? "disabled" : undefined}
                    selected={selectedIndex === idx}
                    onClick={event => this.handleListItemClick(event, idx)} >
                    <ListItemText primary={source.title} />
                </ListItem>
            ));
        return (
            <List aria-label="streams">
                {sourceList}
            </List>
        );
    }

    handleListItemClick(event: React.MouseEvent<HTMLDivElement, MouseEvent>, index: number) {
        event.preventDefault();
        this.setState({
            selectedIndex: index
        });
        this.props.handleSelectSource(this.props.sources[index]);
    }

}