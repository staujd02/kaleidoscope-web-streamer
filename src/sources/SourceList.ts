const SourceList: Array<Source> = [
    {
        source: "https://earth.nullschool.net/#current/particulates/surface/level/overlay=pm10/stereographic=-89.52,26.53,757/loc=-92.818,-73.479",
        title: "Map displaying particulates dispersion"
    },
    {
        source: "https://earth.nullschool.net/#current/ocean/surface/currents/overlay=currents/stereographic=-89.52,26.53,757/loc=-92.818,-73.479",
        title: "A map of the ocean currents"
    },
].concat(chemistryOverlays()).concat(windOverlays());

function windOverlays(): Array<Source> {
    let sources = [] as Array<Source>;
    ["1000hPa", "850hPa", "700hPa", "500hPa", "250hPa", "10hPa"].forEach((pressure: string) => {
        sources.push({
            source: `https://earth.nullschool.net/#current/wind/isobaric/${pressure}/stereographic=-89.52,26.53,757/loc=-92.818,-73.479`,
            title: `A map of wind patterns at ${pressure} pressure`
        });
    });
    return sources;
}

function chemistryOverlays(): Array<Source> {
    let sources = [] as Array<Source>;
    ["cosc", "co2sc", "so2smass"].forEach((chemical: string) => {
        sources.push({
            source: `https://earth.nullschool.net/#current/chem/surface/level/overlay=${chemical}/stereographic=-89.52,26.53,757/loc=-92.818,-73.479`,
            title: `A map of ${chemical} dispertions`
        });
    });
    return sources;
}

export default SourceList;