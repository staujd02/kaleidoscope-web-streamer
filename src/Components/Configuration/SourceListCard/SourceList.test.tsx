import React from 'react';
import { ShallowWrapper, shallow } from "enzyme";
import { formatHTML } from "../../../TestUtilities/htmlFormatter";
import SourceList from "./SourceList";

describe('The Source List', () => {

    let wrapper: ShallowWrapper<SourceListProps, SourceListState, SourceList>
    let props: SourceListProps;

    let handleSelectSource: jest.Mock;

    beforeEach(() => {
        props = {
            handleSelectSource: (handleSelectSource = jest.fn()),
            sources: getSources()
        }
        wrapper = shallow(<SourceList {...props} />);
    });

    it('renders correctly', () => expect(formatHTML(wrapper.html())).toMatchSnapshot());

    describe('when the user searches something', () => {

        beforeEach(() => {
            wrapper.setProps({
                search: '2'
            });
        });

        it('renders correctly', () => expect(formatHTML(wrapper.html())).toMatchSnapshot());
        
        it('narrows the list to the matching elements by url and title', () => {
            expect( wrapper.find('WithStyles(ForwardRef(ListItem))').length)
            .toEqual(2);
        });
    });

    describe('when a user clicks on a list item', () => {

        beforeEach(() => {
            wrapper.find('.disabled').first().simulate('click', {
                preventDefault: jest.fn()
            });
        });

        it('informs the parent element that the item has been selected', () => {
            expect(handleSelectSource).toHaveBeenCalledWith(getSources()[2]);
        });

    });

    function getSources(): Source[] {
        return [
            {
                duration: 6000,
                isEnabled: true,
                sortOrder: 2,
                source: "http:",
                key: 2,
                title: "Title 2"
            },
            {
                duration: 6000,
                isEnabled: true,
                sortOrder: 1,
                source: "http:2",
                key: 1,
                title: "Title 1"
            },
            {
                duration: 6000,
                isEnabled: false,
                sortOrder: 3,
                source: "http",
                key: 3,
                title: "Title 3"
            }
        ]

    }
});