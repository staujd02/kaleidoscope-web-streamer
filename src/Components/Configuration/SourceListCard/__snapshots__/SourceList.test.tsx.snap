// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Source List renders correctly 1`] = `
"
<ul class=\\"MuiList-root MuiList-padding\\" aria-label=\\"streams\\">
  <div class=\\"MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">
    <div class=\\"MuiListItemText-root\\">
      <span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1\\">Title 1</span>
    </div>
  </div>
  <div class=\\"MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">
    <div class=\\"MuiListItemText-root\\">
      <span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1\\">Title 2</span>
    </div>
  </div>
  <div class=\\"MuiButtonBase-root MuiListItem-root disabled MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">
    <div class=\\"MuiListItemText-root\\">
      <span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1\\">Title 3</span>
    </div>
  </div>
</ul>
"
`;

exports[`The Source List when the user searches something renders correctly 1`] = `
"
<ul class=\\"MuiList-root MuiList-padding\\" aria-label=\\"streams\\">
  <div class=\\"MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">
    <div class=\\"MuiListItemText-root\\">
      <span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1\\">Title 1</span>
    </div>
  </div>
  <div class=\\"MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">
    <div class=\\"MuiListItemText-root\\">
      <span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1\\">Title 2</span>
    </div>
  </div>
</ul>
"
`;
